### DISCUSSION TITLE
before_to_disk callback is not invoked

### DISCUSSION THREAD

**Discussion by shaulz** [explosion/spaCy#9241] [Help: Coding & Implementations] [answered]
1. set 
before_to_disk = before_to_disk
in the configuration file 
The configuration file is provided in the attached zip.

2. run
c:\Users\Szevin\PycharmProjects\vaes\venv\Scripts\python.exe -m spacy train C:/projects/IE/services/product-360/DEV_2021_CLAIRE/server.ai/modules/resources/brand/config_S1.cfg --output C:/projects/IE/services/product-360/DEV_2021_CLAIRE/server.ai/models/WDC_English10000 --paths.train ../spaCy_data/WDC_English_10000_train.spacy --paths.dev ../spaCy_data/WDC_English_10000_dev.spacy --code C:/projects/IE/services/product-360/DEV_2021_CLAIRE/server.ai/modules/src/SpacyFunctions.py

The "inside callback" is not printed.

The python module SpacyFunctions.py (attached in zip) includes the callback definition.

* Operating System: windows 10
* Python Version Used: 3.6
* spaCy Version Used: 3.1.2
[SpacyFunctions.zip](https://github.com/explosion/spaCy/files/7168096/SpacyFunctions.zip)

* Environment Information:

**Comment by polm** [comment]
Pasting your code here rather than in a zip file...

```
import spacy

@spacy.registry.callbacks("before_to_disk")
def create_callback():
    def before_to_disk(nlp):
        print("inside callback")
        return nlp

    return before_to_disk
```

<details>

```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "en"
pipeline = ["tok2vec","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
incorrect_spans_key = null
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = before_to_disk

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0
ents_per_type = null

[pretraining]

[initialize]
vectors = "en_core_web_lg"
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```
</details>

**Comment by polm** [comment, marked as answer]
This is a little hard to understand but the right way to register a callback is like this:

```
before_to_disk = {"@callbacks": "my_function"}
```

Using just the function name like you did gives me an error that "before_to_disk is not callable". 

On the other hand I could not produce a case where things ran without issue but the function was not called. Could you give the output of `pip freeze` in case it's something to do with a specific version of a dependency?

**Reply by shaulz** [reply]
Thank you @polm !!! That works fine and callback is invoked.
If you set before_to_disk = null, but still point to SpacyFunctions.py with --code, the code will run fine, but the callback will not be invoked.

**Reply by polm** [reply]
> If you set before_to_disk = null, but still point to SpacyFunctions.py with --code, the code will run fine, but the callback will not be invoked.

That's expected behavior - `--code` just makes the code available, it doesn't necessarily run anything, no matter what the function is named. (I would recommend against naming functions `before_to_disk` in the registry, as the key and value won't be automatically connected and it'll get confusing.)
